substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/events/EndChunkNativePeriodicEvents.java

```diff
@@ -66,14 +66,21 @@ private static void emitInitialEnvironmentVariables(StringEntry[] envs) {
         if (SubstrateJVM.isRecording() && SubstrateJVM.get().isEnabled(JfrEvent.InitialEnvironmentVariable)) {
             JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
             JfrNativeEventWriterDataAccess.initializeThreadLocalNativeBuffer(data);
-
+            long eventId = JfrEvent.InitialEnvironmentVariable.getId();
+            long ticks = JfrTicks.elapsedTicks();
+            int eventIdSize = JfrNativeEventWriter.getEncodedLongSize(eventId);
+            int ticksSize = JfrNativeEventWriter.getEncodedLongSize(ticks);
             for (StringEntry env : envs) {
-                JfrNativeEventWriter.beginEventWrite(data, false);
-                JfrNativeEventWriter.putLong(data, JfrEvent.InitialEnvironmentVariable.getId());
-                JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
+                int encodedSize = JfrNativeEventWriter.getEncodedStringLength(env.key) +
```

Comment:
I don't think that we should compute the size manually as this is error prone and slow. If I recall correctly, HotSpot uses a retry mechanism instead. So, I would suggest to do something similar:
- use `JfrNativeEventSetting.large` to decide if an event should be emitted as a small or as a large event.
- if we tried to emit the event as a small event, but if it turns out that it needs too much space: set `JfrNativeEventSetting.large` to `true` and retry emitting the event. All future events of this event type will then be emitted as large events.

Reply:
Okay. I reduce the scope of this PR, only address the problem on emitting large event. Will take a look retry mechanism in hotspot.

Thanks.